File: unitree_sdk2-py_learn/helloworld/src/publisher.py
import sys
from pathlib import Path
# 获取项目根目录（src的父目录）
root_dir = Path(__file__).parent.parent
# 将根目录添加到系统路径，确保能找到 messages 文件夹
sys.path.append(str(root_dir))

# 之后正常导入
from messages.user_data import *

import time
# 直接把源码文件中unitree_sdk2_python文件夹
# 复制到"~/miniconda3/envs/ros2_humble/lib/python3.10/site-packages"下,否则报警告且无代码提示
import unitree_sdk2py
from unitree_sdk2py.core.channel import ChannelPublisher, ChannelFactoryInitialize

def main():
    print("publisher脚本启动!")

    # 假设从命令行获取输入参数（类似C++的argv[1]）
    if len(sys.argv) != 2:
        print("错误:请传入一个参数!使用方法:python publisher.py <字符串>")
        return
    input_arg = sys.argv[1]
    print(f"您输入的参数是：{input_arg}")

    ChannelFactoryInitialize()

    # Create a publisher to publish the data defined in UserData class
    pub = ChannelPublisher("topic", UserDefineData)
    pub.Init()

    for i in range(50):
        # Create a Userdata message
        msg = UserDefineData(" ", 0)
        msg.string_data = f"HelloWorld! --{input_arg}"
        msg.float_data = time.time()

        # Publish message
        if pub.Write(msg, 0.5):
            print("Publish success. msg:", msg)
        else:
            print("Waitting for subscriber.")

        time.sleep(1)

    pub.Close()

if __name__ == "__main__":
    main()

File: unitree_sdk2-cpp_py_learn/cpp_publisher/publisher.cpp
#include <iostream>
#include <chrono>
#include <thread>
#include "unitree/robot/channel/channel_publisher.hpp"
#include <unitree/common/time/time_tool.hpp>
#include "idl_files/user_data.hpp"  // 由 IDL 文件生成的头文件（需通过 SDK 工具编译）

#define TOPIC "topic_cpp"

using namespace unitree::robot;
using namespace unitree::common;

int main(int argc,char* argv[]) 
{
    std::cout << "cpp_Publisher started." << std::endl;

    //argc:参数数量，argv:参数数组
    // 检查参数数量：argc应为2（程序名 + 1个参数）
    if (argc != 2) {
        std::cerr << "错误：请传入且仅传入一个参数！" << std::endl;
        std::cerr << "使用方法：" << argv[0] << " <字符串参数>" << std::endl;
        return 1; // 返回非0表示程序异常退出
    }
    // 将参数转换为string类型，打印处理后的参数
    std::string input_arg = argv[1];
    std::cout << "您输入的参数是：" << input_arg << std::endl;

    ChannelFactory::Instance()->Init(0);
    ChannelPublisher<UserDataModule::UserData> publisher(TOPIC); // UserDataModule是IDL 中定义的模块名

    publisher.InitChannel();

    // 正确拼接字符串：先将第一个字符串转为std::string再拼接
    std::string message = std::string("cpp_Publisher: ") + "HelloWorld! --" + input_arg;

    while (true)
    {
        UserDataModule::UserData user_data(message.c_str(),static_cast<double>(unitree::common::GetCurrentTimeMillisecond()));;
        publisher.Write(user_data);
        
        // std::cout << "cpp_Publisher Publish success. msg:" << user_data.float_data() <<user_data.string_data() << std::endl;
        if (publisher.Write(user_data,0.5)){
            std::cout << "cpp_Publisher Publish success. msg:" << user_data.float_data() <<user_data.string_data() << std::endl;
        }else{
            std::cout << "cpp_Publisher Waitting for subscriber." << std::endl;
        }
        
        sleep(1);
    }

    // 实际不会执行到这里，仅作为规范
    publisher.CloseChannel();
    return 0;
}